---
title: "TempusLab Bioinformatics Assessment"
author: "Interviewee Xinyi Yan"
date: "`r Sys.Date()`"
output: html_document
---



```{r}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE,message=FALSE)
```

Loading libraries
```{r libraries, include=FALSE, warning=FALSE}
library(vcfR)
library(dplyr)
library(httr)
library(jsonlite)

```

Import vcf data and transform it into a data frame for analysis
```{r load-data, warning=FALSE, message=FALSE}
vcf_data<-read.vcfR('/Users/xinyiyan/Downloads/test_vcf_data.txt')

#Extract all the information and put them into a single data frame, and all the entries in the INFO column are parsed into separate columns
vcf_data_frame<-vcfR2tidy(
  vcf_data,
  info_only = FALSE,
  single_frame = TRUE,
)

vcf_dat<-vcf_data_frame$dat
```

Annotation
```{r,warning=FALSE, message=FALSE}
#Make a new data frame to store the annotations
annotations <- data.frame(as.numeric(vcf_dat$gt_NR)) #The depth of sequence coverage can be annotated using NR in the FORMAT field
colnames(annotations) <- 'DP'
annotations$AD <- as.numeric(vcf_dat$gt_NV) #The number of reads supporting the variant can be annotated using NV in the FORMAT field
annotations$PC <- annotations$AD*100/annotations$DP #Percentage of reads supporting the variants vs. reads supporting the reference reads
       
```


Use the VEP hgvs API to get information 
```{r,warning=FALSE, message=FALSE}
ensembl_vep_url <- 'https://rest.ensembl.org/vep/human/hgvs/'

#Function to query the VEP API for each variant 
annotate_variant_with_vep <- function(chrom, pos, ref, alt) {
  # Build the API URL
  api_endpoint <- paste0(ensembl_vep_url, chrom, ':g.', pos, ref, '>', alt, '?')
  # Make the API request
  response <- GET(url = api_endpoint, add_headers('Content-Type' = 'application/json'))
  vep_data <- fromJSON(content(response, 'text', encoding = 'UTF-8')) 
  
  # Initialize annotations with default values
  gene <- 'NA'
  variant_type <- 'NA'
  effect <- 'NA'
  minor_allele_frequency <- 'NA'
  
  # Check if the API request was successful and information is available
  if (length(vep_data) > 1) {
    if (!is.null(vep_data$transcript_consequences[[1]]$gene_symbol)) {
      gene <- vep_data$transcript_consequences[[1]]$gene_symbol[1]
    }
    if (!is.null(vep_data$most_severe_consequence)) {
      variant_type <- vep_data$most_severe_consequence
    }
    if (!is.null(vep_data$transcript_consequences[[1]]$impact)) {
      effect <- vep_data$transcript_consequences[[1]]$impact[1]
    } else if (!is.null(vep_data$intergenic_consequences[[1]]$impact)) {
      effect <- vep_data$intergenic_consequences[[1]]$impact[1]
    }
    if (!is.null(vep_data$colocated_variants[[1]]$gnomad_af)) {
      minor_allele_frequency <- vep_data$colocated_variants[[1]]$gnomad_af
    }
  }
  
  # Return the annotation information
  return(list(gene = gene, variant_type = variant_type, effect = effect, minor_allele_frequency = minor_allele_frequency))
}



#To save some running time, I only annotated the first 2000 variants
dat_demo<- vcf_dat[1:1000, ]
annotations_demo<- annotations[1:1000, ]

#Annotate each variant 
for (i in 1:nrow(dat_demo)) {
  annotation <- annotate_variant_with_vep(dat_demo$CHROM[i], 
                                          dat_demo$POS[i], 
                                          dat_demo$REF[i], 
                                          dat_demo$ALT[i])
 annotations_demo$gene[i] <- annotation$gene
 annotations_demo$variant_type[i] <- annotation$variant_type
 annotations_demo$effect[i]<- annotation$effect
 annotations_demo$minor_allele_frequency[i]<-annotation$minor_allele_frequency
}


#Save the annotations into a csv file
write.csv(annotations_demo, "/Users/xinyiyan/Desktop/variant_annotations.csv", row.names = FALSE)

```

